datalog/datalog.go

```diff
@@ -630,6 +639,50 @@ func (t *SymbolTable) Str(sym Symbol) string {
 	return (*t)[int(sym)]
 }
 
+func (t *SymbolTable) Clone() *SymbolTable {
+	newTable := *t
+	return &newTable
+}
+
+// SplitOff returns a newly allocated slice containing the elements in the range
+// [at, len). After the call, the receiver will be left containing
+// the elements [0, at) with its previous capacity unchanged.
+func (t *SymbolTable) SplitOff(at int) *SymbolTable {
+	if at > len(*t) {
+		panic("split index out of bound")
+	}
+
+	new := (*t)[at:]
+	*t = (*t)[:at]
```

Comment:
Is there any scenario where the returned `SymbolTable` and this "parent" will be both appended to later? If so, we need to copy as we could inadvertently modify another table. If not, add a comment noting that the returned value is intended to be immutable.